### PR TITLE
[WIP] make folder structure and makefile

### PR DESCRIPTION
**Pull request by SteveBronder**
This is going to be a bit of a longer PR I think.

With all the focus on performance lately I'd like there to be some level of structure. So when people say, "Hey I'm seeing a __% increase" it's easy for them to show that with some tests.

This PR is going to

- [ ] Setup the makefile for individual tests
- [ ] ./runTests.py so that users can make or checkout and test one math PR against another with a set of tests
- [ ] Add a PR template (?)

My overall goal here would be so that if someone makes a `Performance Tests for PR #1141Blah` then the Jenkins in the math repo can read through the PRs here, run the performance tests against develop, then post a message with the results to the math PR. This feels like a nice solution, where issues that are non-performance based can not make a PR here, but ones that make performance assertions will have a standard template to show those against mac/linux/windows builds.

### CODE REVIEW

**Comment by SteveBronder** [comment]
Oh, Sean this also has performance tests for the two vec_concat methods, but the tl;dr there is that the performance difference is negligible and less code is better to me so yours is the better implementation

**Comment by SteveBronder** [comment]
Should I writeup a technical spec for this? Think it may be good to try to do that, then post to discourse for discussion for interested parties

**Comment by syclik** [comment]
Yes, please. We can refine that so we tie into our continuous integration
process. I've been thinking about it, but haven't taken action yet. Thanks!

**Comment by SteveBronder** [comment]
> I've been thinking about it, but haven't taken action yet. Thanks!

Np! I think we've all been thinking about it, mostly wanted to setup a PR to get the ball rolling. I have some other things I have to do this weekend but can start this next weekend

**Comment by seantalts** [comment]
Thank you! Will take a look at this PR. Re: the Jenkins integration, I think that's something really important. Though I don't really think we need to necessarily automate any cross-linking between a Math PR and a PR here, other than that a user might want to run the Jenkins job for this repo on certain microbenchmarks that they or the reviewer find informative against the full suite of OSes. 

Separately, I'm asking our new dev ops guy to look at integrating a few shorter end to end performance tests (stat comp benchmark models, likely) into Math and Stan PRs soon just to test for performance regressions.

File: tests/opencl/.gitignore [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+!/test/*/*/*/*/
+!/test/*/*/*/
+!/test/*/*/
+!/test/*/
+!/test/*/*/*/*/*/*.*
+!/test/*/*/*/*/*.*
+!/test/*/*/*/*.*
+!/test/*/*/*.*
+!/test/*/*.*
+!/test/*.*
```
Comment (seantalts) [inline tests/opencl/.gitignore:19]: What is going on here?

File: .gitignore [hunk 2]
```diff
@@ -21,7 +21,7 @@
 /test/**/*.xml
 /test/dummy.cpp
 /test/**/*.so
-
+**/math/**/*
```
Comment (seantalts) [inline .gitignore:24]: Doesn't this ignore any math directories anywhere? I also don't think we want to ignore the math submodule in general. What was this aimed at?

File: tests/opencl/vec_cat/vec_cat_event_test.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  std::vector<cl::Event> v2;
+  std::vector<cl::Event> v3;
+  cl::Event test_event;
+
+  for (auto _ : state) {
+    state.PauseTiming();
+    benchmark::DoNotOptimize(test_event);
+    // A lot of the time we'll have an empty matrix with other two having
+    // An extra dependency
+    if (state.range(0) == 0) {
```
Comment (seantalts) [inline tests/opencl/vec_cat/vec_cat_event_test.cpp:86]: Can you do this setup outside of this for loop?

**Review by seantalts — CHANGES_REQUESTED** [review]
Thanks for taking a crack at this! Much needed reorg and some great new benchmarks.

re directories - 3 of the ones you have in `examples` are actually about eigen stuff, I'd just lump all of that together. None of them are specifically "examples" - they all answer specific questions I had, haha.

is this a WIP PR with more stuff being added? If it's okay with you I'd like to just merge the current scope (after comments addressed) and then do the longer list with separate PRs if that's okay? Just because I already looked over this one.

File: tests/opencl/.gitignore (hunk 1 above)
Comment (SteveBronder) [reply tests/opencl/.gitignore:19]: Whups!

File: .gitignore (hunk 2 above)
Comment (SteveBronder) [reply .gitignore:24]: Oh I thought we would ignore the math repo and users would have it locally. I can remove this

File: tests/opencl/vec_cat/vec_cat_event_test.cpp (hunk 3 above)
Comment (SteveBronder) [reply tests/opencl/vec_cat/vec_cat_event_test.cpp:86]: Yep!

**Comment by SteveBronder** [comment]
ftr this is def a work in progress PR. I'm pro making the scope of this what's already here. Next week I'll have more time to fix this up and can flag you for review on it

**Comment by seantalts** [comment]
ah word. is there some way to mark it as WIP? I saw that on another PR in another repo once... not the [WIP] in title but some github mechanism.

**Comment by SteveBronder** [comment]
I can make a label? Idk tho

**Pull request closed without merging**